Finished with oop exercise

kalash14 opened the pull request:
_Complete the form below and remove this line._

### OOP exercise

[Demo](https://kalash14.github.io/oop-exercise/) |
[Code base](https://github.com/kalash14/oop-exercise/blob/master/index.js)

The code is submitted in a dedicated feature branch.
Please, review.

OleksiyRudenko commented:
@kalash14 Looks good at a glance. Please, meanwhile check yourself against the following list:

 * [ ] Implement a base class to inherit from
 * [ ] Employ default parameters
 * [ ] Each species is represented with its own class
 * [ ] No need to specify species at instantiation
 * [ ] Classes for species that do not have hands by natural design
       do not consequently have `hands` or any equivalent property
 * [ ] All inhabitants are stored in a container (array or object)
 * [ ] JS native features are intensively employed (`const`, `let`, `Array.map|join|forEach|...`, etc)
 * [ ] [OOP, SOLID and DRY](https://github.com/OleksiyRudenko/a-tiny-JS-world/blob/master/README.md#learn-on-your-own)
       principles are intensively employed
 
Optional level up (not required to implement):
 * [ ] Friends list is a list of objects refs rather than names (strings)
 * [ ] Cat-woman class is built employing composition rather than inheritance only
 
[ES6 classes cheat-sheet](https://gist.github.com/OleksiyRudenko/672d39b08d9d0da4e179aca49876c58b)

Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

OleksiyRudenko commented:
@kalash14 Please, let me know somehow that you have implemented everything you wanted to and also 
> Please, confirm all items above are positive. Please, ask a question if anything above requires clarification.

kalash14 commented:
In my previous commit, I have reworked my code according to your recommendations and I have implemented the next items:

- [x] Implement a base class to inherit from
- [x] Employ default parameters
- [x] Each species is represented with its own class
- [x] No need to specify species at instantiation
- [x] Classes for species that do not have hands by natural design
do not consequently have hands or any equivalent property
- [x] All inhabitants are stored in a container (array or object)
- [x] JS native features are intensively employed (const, let, Array.map|join|forEach|..., etc)
- [x] OOP, SOLID and DRY principles are intensively employed 

AFAIU, I should make a commit after every implemented item from the list above. Right? Thanks for your answer.

kalash14 commented:
Also, to my mind, in order to follow DRY principles, I would like to create an instance 'parameters' and put there all parameters of the base class, because I repeated them in my code and I think it's not OK

OleksiyRudenko commented:
> AFAIU, I should make a commit after every implemented item from the list above. Right? Thanks for your answer.

It is up to you. Just let me know when you finish with code improvements or have a question.

OleksiyRudenko commented:
> Also, to my mind, in order to follow DRY principles, I would like to create an instance 'parameters' and put there all parameters of the base class, because I repeated them in my code and I think it's not OK

Good catch ;)

Whatever makes code DRY is good.

kalash14 commented:
I have finished with code improvements.

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,86 @@
+
+class Human extends Inhabitant {
+
+    constructor(name, gender, greetingsWord, kind, favouriteDish, legs = 2) {
```
Do you really want to let the caller to provide `Human` with any kind at caller's discretion?

kalash14 replied:
Oops...

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,86 @@
+}
+
+class Dog extends Animal {
+    constructor(name, gender, greetingsWord = 'Bow-wow', kind = 'dog', favouriteDish, legs) {
```
Do you really want to let the caller to provide `Dog` with any kind at caller's discretion?
Do you really want to let the caller to provide `Dog` with any number of legs at caller's discretion?

Ask same question on every parameter of every class constructors. Respond wisely and fix correspondingly.

kalash14 replied:
Ok

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,86 @@
+
+    new Dog('Sharik', 'male', 'Bow-wow', 'dog', 'meat'),
+    new Cat('Kuzia', 'male', 'Meow-meow', 'cat', 'boiled fish'),
+    new Woman('Leila', 'female', 'Hi', 'woman', 'Cesar salad'),
```
Please, note that `woman` and `man` are both of human species.

kalash14 replied:
Ok, thanks

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,86 @@
+    new Man('Pedro Rodriguez', 'male', 'Hey, dude', 'man', 'Avocado sandwich')
+
+];
+
```
Please, read carefully docs with regards to output and fix the code below correspondingly.

kalash14 replied:
Ok

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,69 @@
+
+class Inhabitant {
+
+    constructor(name, gender) {
```
Please, initialize in `Inhabitant` all the properties that are common for all inhabitants.

kalash14 replied:
Common properties for all inhabitants: name, gender, kind and greetingsWords

OleksiyRudenko replied:
@kalash14 That's correct. Please, fix `Inhabitant`.

kalash14 replied:
Ok

OleksiyRudenko commented:
@kalash14 Well, I hoped we can complete the task in 3 steps.
Please, compare your own suggestion 
> Common properties for all inhabitants are name, gender, kind, greetingsWords and legs
with your current code.

Please, take comments as verbose technical requirements documentation.

kalash14 commented:
I noticed that I didn't inherit these properties in each class

kalash14 commented:
Thank you for your patience and sorry for my inattention. I have fixed these parameters.

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+
+class Human extends Inhabitant{
+
+    constructor(name, gender, greetingsWords, legs = 2) {
```
You can ask user for hands as an argument to `new Human` with a default value like `legs`.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+    constructor(name, gender, greetingsWords, legs = 2) {
+        super(name, gender, 'Human', greetingsWords, legs);
+        this.hands = 2;
+        this.greetingsWords = `Hi!`
```
Why do you override `greetingsWords`?
You receive something from a user and pass to the base class via `super`, so the property gets initialized properly.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+
+class Dog extends Animal {
+
+    constructor(name, gender, greetingsWords, legs = 4) {
```
If you want a dog to have a standard greeting then you do need to ask a user for one.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+    constructor(name, gender, greetingsWords, legs = 4) {
+        super(name, gender, 'Dog', greetingsWords, legs);
+        this.greetingsWords = 'Bow-wow!';
+    }
```
And when calling `super` you can specify the default greeting explicitly.
Super class will set the property for you.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+    constructor(name, gender, greetingsWords, legs = 4) {
+        super(name, gender, 'Cat', greetingsWords, legs);
+        this.greetingsWords = 'Meow!';
+    }
```
Same as for `Dog`.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+const creatures = [
+
+    new Human('Victor', 'male'),
+    new Human('Helen', 'female'),
```
Let humans have their individual greetings.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,66 @@
+
+];
+
+creatures.forEach(creature => print(creature.kind + "; " + creature.gender + "; " + creature.name + "; " + creature.legs + "; " + (creature.hands ? creature.hands + "; " : "") + creature.greetingsWords));
```
Please, add to the basemost class a method that would return a nicely formatted string. That's how separation of concerns principle gets employed.

Then here run through array and call that method for every creature within `print()`.

kalash14 replied:
ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,68 @@
+        this.hands = hands;
+    }
+
+}
```
In method overloading we trust.
Define `greetingMessage` for `Human`. Use `greetingMessage` from super class (`super.greetingMessage()`) and add something about hands. No problem that hands come after everything else.

BTW, as you might remember function names should start with a verb.

As you have to rename it anyway rename `greetingMessage` to `toString` everywhere. Check for the bonus learning below.

kalash14 replied:
Ok!

OleksiyRudenko commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,68 @@
+
+];
+
+creatures.forEach(creature => print(creature.greetingMessage()));
```
If your class (object) has magic method `toString` you can benefit from it in string contexts.
`print(creature)` will call `.toString()` implicitly for you.

OleksiyRudenko replied:
Please, ensure your files have an empty line at the end. It's important. You can tune your IDE so it ensures empty lines for you.

kalash14 replied:
Wow! Cool! 
Ok, I have already configured my IDE.

kalash14 commented on submissions/kalash14/oop-exercise/index.js:
```diff
@@ -0,0 +1,68 @@
+
+    greetingMessage() {
+        return `${this.kind}; ${this.gender}; ${this.name}; ${this.legs}; ${(this.hands ? this.hands + "; " : "")}${this.greetingsWords}`;
+    }
```
Ok!

kalash14 commented:
Thanks for all recommendations! Happy New year and Merry Christmas!!!
